### PR TITLE
add cache for everything

### PR DESCRIPTION
**Pull request by mjudeikis**
Adds cache for cluster object for all clusters. 
For now cache is invalidated only on restart or key authentication failure. 

What if we move to time-based cache? https://github.com/patrickmn/go-cache ? 

/cc kargakis

### CODE REVIEW

File: cap/pkg/clusters/enumerator.go [hunk 1]
```diff
@@ -18,9 +18,14 @@ import (
 	"github.com/Azure/go-autorest/autorest"
 	"github.com/Azure/go-autorest/autorest/azure/auth"
 	"github.com/openshift/openshift-azure/pkg/api"
+	cache "github.com/patrickmn/go-cache"
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: FWIW, we have caching libraries already in kube, maybe worth trying to reuse vs importing a new one.

https://github.com/openshift/azure-misc/blob/master/vendor/k8s.io/client-go/tools/cache/mutation_cache.go

Comment (mjudeikis) [reply cap/pkg/clusters/enumerator.go]: This looks promising. I like go-cache as its super simple, but if it does the job and ones less dep - im in.

**Comment by mjudeikis** [comment]
@kargakis ptal

File: cap/pkg/cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	if !ok {
+		return nil, nil, false
+	}
+	return value.(objectEntry).obj, nil, true
+}
+
+// Add objectEntry by using a unique string as the key.
+func (c *ObjectCache) Add(key string, obj interface{}) error {
+	err := c.cache.Add(objectEntry{key: key, obj: obj})
```
Comment (kargakis) [inline cap/pkg/cache/cache.go]: nit: `return c.cache.Add(objectEntry{key: key, obj: obj})`

File: cap/pkg/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete objectEntry by using a unique string as the key.
+func (c *ObjectCache) Delete(key string, obj interface{}) error {
+	err := c.cache.Delete(objectEntry{key: key, obj: obj})
```
Comment (kargakis) [inline cap/pkg/cache/cache.go]: nit: `reutrn c.cache.Delete(objectEntry{key: key, obj: obj})`

File: cap/pkg/clusters/enumerator.go [hunk 4]
```diff
@@ -157,83 +164,111 @@ func (e *enumerator) getSubscriptionClusters(ctx context.Context, clusters map[s
 }
 
 func (e *enumerator) getSubscriptionCluster(ctx context.Context, subscriptionID, resourceGroup string) (*Cluster, error) {
+
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: kill this newline

File: cap/pkg/clusters/enumerator.go [hunk 5]
```diff
@@ -17,7 +17,10 @@ import (
 	azstorage "github.com/Azure/azure-sdk-for-go/storage"
 	"github.com/Azure/go-autorest/autorest"
 	"github.com/Azure/go-autorest/autorest/azure/auth"
+
+	"github.com/openshift/azure-misc/cap/pkg/cache"
 	"github.com/openshift/openshift-azure/pkg/api"
+
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: We need CI to be running goimports. Mind opening an issue for that? This should be

stdlib

third-party

azure-misc

File: cap/pkg/clusters/enumerator.go [hunk 6]
```diff
@@ -32,18 +35,22 @@ type enumerator struct {
 	exclude *regexp.Regexp
 
 	clusters map[string]Cluster
+
+	store cache.ObjectCache
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: comment that mentions that this is a resource group cache

File: cap/pkg/clusters/enumerator.go [hunk 7]
```diff
@@ -32,18 +35,22 @@ type enumerator struct {
...
 }
 
+var cacheTTL = 30 * time.Minute
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: seems arbitrary, why 30m? at least a comment would be helpful

Comment (mjudeikis) [reply cap/pkg/clusters/enumerator.go]: because it is arbitrary. finger in the air constant :)

**Comment by mjudeikis** [comment]
@kargakis done. will fix conflict later before merge

File: cap/pkg/clusters/enumerator.go [hunk 8]
```diff
@@ -157,83 +165,110 @@ func (e *enumerator) getSubscriptionClusters(ctx context.Context, clusters map[s
...
+		//populate cache
+		e.store.Add(fmt.Sprintf("%s/%s", subscriptionID, resourceGroup), cluster)
+		return &Cluster{
+			Name:        resourceGroup,
+			WorkspaceID: *(*workspaces.Value)[0].CustomerID,
+			Manifest:    cs,
+		}, nil
+	}
+	e.log.Debug("return result from cache")
+	cluster, err, found := e.store.Get(fmt.Sprintf("%s/%s", subscriptionID, resourceGroup))
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: You already get this once above, you probably don't need this.

File: cap/pkg/clusters/enumerator.go [hunk 9]
```diff
@@ -157,83 +165,110 @@ func (e *enumerator) getSubscriptionClusters(ctx context.Context, clusters map[s
...
 	if err != nil {
 		return nil, err
 	}
-	if len(*workspaces.Value) != 1 {
-		return nil, fmt.Errorf("found %d workspaces", len(*workspaces.Value))
+	if !found && err == nil {
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: At this point found should be true so you can remove this.

File: cap/pkg/clusters/enumerator.go [hunk 10]
```diff
@@ -157,83 +165,110 @@ func (e *enumerator) getSubscriptionClusters(ctx context.Context, clusters map[s
...
 	if err != nil {
 		return nil, err
 	}
-
-	var acct *storage.Account
-	for i := range *accts.Value {
-		if (*accts.Value)[i].Tags["type"] != nil && *(*accts.Value)[i].Tags["type"] == "config" {
-			acct = &(*accts.Value)[i]
-			break
+	if !found && err == nil {
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: err will always be nil here so this can be trimmed down to `if !found {`

File: cap/pkg/clusters/enumerator.go [hunk 11]
```diff
@@ -157,83 +165,110 @@ func (e *enumerator) getSubscriptionClusters(ctx context.Context, clusters map[s
 }
 
 func (e *enumerator) getSubscriptionCluster(ctx context.Context, subscriptionID, resourceGroup string) (*Cluster, error) {
+	var acct *storage.Account
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: Why this change? You seem to use it only when you are doing the live lookup.

File: cap/pkg/clusters/enumerator.go [hunk 12]
```diff
@@ -157,83 +165,110 @@ func (e *enumerator) getSubscriptionClusters(ctx context.Context, clusters map[s
...
+		if len(*workspaces.Value) != 1 {
+			return nil, fmt.Errorf("found %d workspaces", len(*workspaces.Value))
+		}
+
+		cluster := Cluster{
+			Name:        resourceGroup,
+			WorkspaceID: *(*workspaces.Value)[0].CustomerID,
+			Manifest:    cs,
+		}
+		//populate cache
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: nit: add a space between // and the start of the comment

File: cap/pkg/clusters/enumerator.go (hunk 11 above)
Comment (mjudeikis) [reply cap/pkg/clusters/enumerator.go]: because its being assigned in the loop, and its not existing outside. Will move lower to reduce scope

File: cap/pkg/clusters/enumerator.go [hunk 13]
```diff
@@ -160,80 +168,99 @@ func (e *enumerator) getSubscriptionCluster(ctx context.Context, subscriptionID,
...
 	stc.Authorizer = e.authorizer
 
-	accts, err := stc.ListByResourceGroup(ctx, resourceGroup)
+	//if we dont have cached key - get one
+	cluster, err, found := e.store.Get(fmt.Sprintf("%s/%s", subscriptionID, resourceGroup))
 	if err != nil {
 		return nil, err
 	}
-
-	var acct *storage.Account
```
Comment (kargakis) [inline cap/pkg/clusters/enumerator.go]: revert this change

Comment (mjudeikis) [reply cap/pkg/clusters/enumerator.go]: why so?

Comment (kargakis) [reply cap/pkg/clusters/enumerator.go]: it's 1. unnecessary and 2. bad practice to not keep variables as close to where they are used.

Comment (mjudeikis) [reply cap/pkg/clusters/enumerator.go]: I need it in the loop to check acc availability, so I need it. For closines, I can move it like 5 lines down to be closer to `range` I used it in first.
